### ISSUE TITLE
[Question] PCI Devices Pod Restarts

### ISSUE THREAD

**Issue by schmoaaaaah** [harvester/harvester#3871]
I have a 3 node cluster. Freshly installed. I enabled the PCI Devices Addon and it worked fine on two nodes, but on one it fails and restarts constantly with the following log.
```
time="2023-05-08T08:44:02Z" level=info msg="Applying CRD pcidevices.devices.harvesterhci.io"
time="2023-05-08T08:44:02Z" level=info msg="Applying CRD pcideviceclaims.devices.harvesterhci.io"
time="2023-05-08T08:44:11Z" level=info msg="Registering PCI Device Claims controller"
time="2023-05-08T08:44:12Z" level=info msg="Loading driver vfio-pci"
time="2023-05-08T08:44:12Z" level=info msg="Loading driver vfio_iommu_type1"
time="2023-05-08T08:44:12Z" level=info msg="Registering PCI Devices controller"
time="2023-05-08T08:44:12Z" level=info msg="add mutation handler for [pods]. (Pod)"
time="2023-05-08T08:44:12Z" level=info msg="add mutation handler for [virtualmachines].kubevirt.io (VirtualMachine)"
time="2023-05-08T08:44:12Z" level=info msg="Active TLS secret harvester-system/pcidevices-webhook-tls (ver=25321) (count 1): map[listener.cattle.io/cn-pcidevices-webhook.harvester-system.svc:pcidevices-webhook.harvester-system.svc listener.cattle.io/fingerprint:SHA1=06E69757A946BC22422ACA20E153E1DBBCBCF903]"
time="2023-05-08T08:44:12Z" level=info msg="Listening on :8443"
I0508 08:44:13.628762       1 trace.go:205] Trace[911902081]: "DeltaFIFO Pop Process" ID:harv3-0000ff0c0,Depth:30,Reason:slow event handlers blocking the queue (08-May-2023 08:44:13.430) (total time: 197ms):
Trace[911902081]: [197.83893ms] [197.83893ms] END
time="2023-05-08T08:44:20Z" level=info msg="Starting /v1, Kind=Node controller"
time="2023-05-08T08:44:20Z" level=info msg="Starting network.harvesterhci.io/v1beta1, Kind=VlanConfig controller"
time="2023-05-08T08:44:20Z" level=info msg="Starting devices.harvesterhci.io/v1beta1, Kind=PCIDevice controller"
time="2023-05-08T08:44:20Z" level=info msg="Starting devices.harvesterhci.io/v1beta1, Kind=PCIDeviceClaim controller"
time="2023-05-08T08:44:33Z" level=info msg="Reconciling PCI Devices list"
I0508 08:44:42.430276       1 trace.go:205] Trace[646203300]: "Reflector ListAndWatch" name:github.com/harvester/pcidevices/go/pkg/mod/k8s.io/client-go@v0.23.7/tools/cache/reflector.go:167 (08-May-2023 08:44:17.628) (total time: 24801ms):
Trace[646203300]: ---"Objects listed" error:<nil> 24801ms (08:44:42.429)
Trace[646203300]: [24.801374748s] [24.801374748s] END
I0508 08:44:42.628726       1 trace.go:205] Trace[1747278511]: "DeltaFIFO Pop Process" ID:recurringjobs.longhorn.io,Depth:43,Reason:slow event handlers blocking the queue (08-May-2023 08:44:42.528) (total time: 100ms):
Trace[1747278511]: [100.017068ms] [100.017068ms] END
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1344034]

goroutine 31 [running]:
github.com/harvester/pcidevices/pkg/util/nichelper.IdentifyHarvesterManagedNIC({0xc00005a00a, 0x5}, {0x1f0c1f0, 0xc001cb64b0}, {0x1f0bc40, 0xc00279bb30})
	/go/src/github.com/harvester/pcidevices/pkg/util/nichelper/helper.go:52 +0x454
github.com/harvester/pcidevices/pkg/controller/pcidevice.Register({0x1f0b968, 0xc002719fc0}, {0x7f1403b36f28, 0xc0004222a0}, 0xc00050c0d8, 0xc00050c0e0)
	/go/src/github.com/harvester/pcidevices/pkg/controller/pcidevice/pcidevice_controller.go:69 +0x345
main.run.func2()
	/go/src/github.com/harvester/pcidevices/main.go:148 +0x72
golang.org/x/sync/errgroup.(*Group).Go.func1()
	/go/src/github.com/harvester/pcidevices/go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
	/go/src/github.com/harvester/pcidevices/go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:72 +0xa5
```
This log repeats with every restart. Only the 
```
I0508 08:44:13.628762       1 trace.go:205] Trace[911902081]: "DeltaFIFO Pop Process" ID:harv3-0000ff0c0,Depth:30,Reason:slow event handlers blocking the queue (08-May-2023 08:44:13.430) (total time: 197ms):
Trace[911902081]: [197.83893ms] [197.83893ms] END
``` 
appears at first start and then never again.
Runnin `lspci` in the pod or on the Host works.

**Comment by schmoaaaaah** [comment]
I tried it with the 0.2.2 image and it worked to find the devices using them however not.
So with the claims created i upgraded to 0.2.3 and the claims worked.

**Comment by w13915984028** [comment]
also cc @ibrokethecloud

**Comment by w13915984028** [comment]
@Schmoaaaaah Is it working smoothly now ? thanks.

**Comment by schmoaaaaah** [comment]
Hi.
Sry for the late answer. I sadly had to go back to proxmox as I got to many Problems with Harvester.
It technically worked after the version switches, but there were still a error displayed in the pci pod and starting the VM that had the PCI Devices attached would tacke hours. Especially the HBA i passed through was a bit buggy and needed some restarts and dis/enabeling the pci addon.

**Comment by mircea-pavel-anton** [comment]
I am having a similar issue running Harvester 1.1.2 as a VM under Proxmox (proof of concept before switching over)

I enabled the pcie devices addon on my VM and the pod constantly restarts. Here are the logs:

```
time="2023-08-28T22:42:19Z" level=info msg="Applying CRD pcidevices.devices.harvesterhci.io"
2023-08-28T22:42:19.477962358Z time="2023-08-28T22:42:19Z" level=info msg="Applying CRD pcideviceclaims.devices.harvesterhci.io"
time="2023-08-28T22:42:25Z" level=info msg="Registering PCI Device Claims controller"
time="2023-08-28T22:42:25Z" level=info msg="Loading driver vfio-pci"
time="2023-08-28T22:42:25Z" level=info msg="Loading driver vfio_iommu_type1"
time="2023-08-28T22:42:25Z" level=info msg="Registering PCI Devices controller"
time="2023-08-28T22:42:25Z" level=info msg="add mutation handler for [pods]. (Pod)"
2023-08-28T22:42:25.733436963Z time="2023-08-28T22:42:25Z" level=info msg="add mutation handler for [virtualmachines].kubevirt.io (VirtualMachine)"
time="2023-08-28T22:42:25Z" level=info msg="Active TLS secret harvester-system/pcidevices-webhook-tls (ver=69607) (count 1): map[listener.cattle.io/cn-pcidevices-webhook.harvester-system.svc:pcidevices-webhook.harvester-system.svc listener.cattle.io/fingerprint:SHA1=A0AA2D9C3ADE6AA81C4173405E0D7C0D9095F1A4]"
time="2023-08-28T22:42:25Z" level=info msg="Listening on :8443"
time="2023-08-28T22:42:28Z" level=info msg="Starting devices.harvesterhci.io/v1beta1, Kind=PCIDeviceClaim controller"
2023-08-28T22:42:28.635538546Z time="2023-08-28T22:42:28Z" level=info msg="Starting /v1, Kind=Node controller"
2023-08-28T22:42:28.635545302Z time="2023-08-28T22:42:28Z" level=info msg="Starting network.harvesterhci.io/v1beta1, Kind=VlanConfig controller"
time="2023-08-28T22:42:28Z" level=info msg="Starting devices.harvesterhci.io/v1beta1, Kind=PCIDevice controller"
I0828 22:42:40.233165       1 trace.go:205] Trace[208240456]: "Reflector ListAndWatch" name:github.com/harvester/pcidevices/go/pkg/mod/k8s.io/client-go@v0.23.7/tools/cache/reflector.go:167 (28-Aug-2023 22:42:28.434) (total time: 11798ms):
2023-08-28T22:42:40.233416888Z Trace[208240456]: ---"Objects listed" error:<nil> 11601ms (22:42:40.036)
2023-08-28T22:42:40.233422781Z Trace[208240456]: [11.798485857s] [11.798485857s] END
time="2023-08-28T22:42:45Z" level=info msg="Reconciling PCI Devices list"
I0828 22:42:47.734084       1 trace.go:205] Trace[1106410694]: "Reflector ListAndWatch" name:github.com/harvester/pcidevices/go/pkg/mod/k8s.io/client-go@v0.23.7/tools/cache/reflector.go:167 (28-Aug-2023 22:42:28.434) (total time: 19299ms):
2023-08-28T22:42:47.734403747Z Trace[1106410694]: ---"Objects listed" error:<nil> 19299ms (22:42:47.733)
2023-08-28T22:42:47.734412370Z Trace[1106410694]: [19.299423006s] [19.299423006s] END
time="2023-08-28T22:42:47Z" level=info msg="Starting kubevirt.io/v1, Kind=VirtualMachine controller"
2023-08-28T22:42:47.834562753Z time="2023-08-28T22:42:47Z" level=info msg="Starting devices.harvesterhci.io/v1beta1, Kind=PCIDevice controller"
2023-08-28T22:42:47.834567737Z time="2023-08-28T22:42:47Z" level=info msg="Starting devices.harvesterhci.io/v1beta1, Kind=PCIDeviceClaim controller"
time="2023-08-28T22:42:47Z" level=info msg="Starting /v1, Kind=Secret controller"
2023-08-28T22:42:47.834707455Z time="2023-08-28T22:42:47Z" level=info msg="Sleeping for 15 seconds then applying webhook config"
2023-08-28T22:42:47.834802399Z time="2023-08-28T22:42:47Z" level=info msg="Starting apiextensions.k8s.io/v1, Kind=CustomResourceDefinition controller"
2023-08-28T22:42:47.835028488Z time="2023-08-28T22:42:47Z" level=info msg="Starting apiregistration.k8s.io/v1, Kind=APIService controller"
time="2023-08-28T22:42:47Z" level=info msg="Updating TLS secret for harvester-system/pcidevices-webhook-tls (count: 1): map[listener.cattle.io/cn-pcidevices-webhook.harvester-system.svc:pcidevices-webhook.harvester-system.svc listener.cattle.io/fingerprint:SHA1=A0AA2D9C3ADE6AA81C4173405E0D7C0D9095F1A4]"
panic: runtime error: invalid memory address or nil pointer dereference
2023-08-28T22:42:50.738339744Z [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1344034]
2023-08-28T22:42:50.738345954Z 
2023-08-28T22:42:50.738350720Z goroutine 88 [running]:
2023-08-28T22:42:50.738356027Z github.com/harvester/pcidevices/pkg/util/nichelper.IdentifyHarvesterManagedNIC({0xc000058022, 0x7}, {0x1f0c1f0, 0xc002ffe8d0}, {0x1f0bc40, 0xc002d9c270})
2023-08-28T22:42:50.738360230Z 	/go/src/github.com/harvester/pcidevices/pkg/util/nichelper/helper.go:52 +0x454
github.com/harvester/pcidevices/pkg/controller/pcidevice.Register({0x1f0b968, 0xc002d0fd80}, {0x7f5f4e4226e8, 0xc0001603f0}, 0xc0006002c0, 0xc0006002d0)
2023-08-28T22:42:50.738383324Z 	/go/src/github.com/harvester/pcidevices/pkg/controller/pcidevice/pcidevice_controller.go:69 +0x345
2023-08-28T22:42:50.738388030Z main.run.func2()
2023-08-28T22:42:50.738393222Z 	/go/src/github.com/harvester/pcidevices/main.go:148 +0x72
golang.org/x/sync/errgroup.(*Group).Go.func1()
2023-08-28T22:42:50.738401983Z 	/go/src/github.com/harvester/pcidevices/go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75 +0x64
2023-08-28T22:42:50.738405691Z created by golang.org/x/sync/errgroup.(*Group).Go
	/go/src/github.com/harvester/pcidevices/go/pkg/mod/golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:72 +0xa5
```

**Comment by ibrokethecloud** [comment]
this has been fixed in 1.2.0. The issue is in harvester is installed in a VM, the default nic type is likely virtio.. and that has no pci address. If you wish to use 1.1.2 may be try using a different virtual nic type.
